File: Source/Gamemodes/Solo/SoloDeathTimer.cs
﻿using WCSharp.Api;
using static WCSharp.Api.Common;

public class SoloDeathTimer
{
    private const float TIME_TO_REVIVE = 6.0f;
    private const float TextTagHeight = 0.018f;
    private const float Y_OFFSET = 5.0f;
    public AchesTimers ReviveTimer;
    public AchesTimers UpdateTextTimer;
    public player Player;
    public texttag FloatingTimer;

    public SoloDeathTimer(player player)
    {
        Player = player;
        ReviveTimer = ObjectPool.GetEmptyObject<AchesTimers>();
        UpdateTextTimer = ObjectPool.GetEmptyObject<AchesTimers>();
        FloatingTimer = CreateFloatingTimer();
        StartTimers();
    }

    private texttag CreateFloatingTimer()
    {
        var circle = Globals.ALL_CIRCLES[Player];
        var floatText = texttag.Create();
        floatText.SetPosition(circle.Unit.X, circle.Unit.Y - Y_OFFSET, 0);
        floatText.SetVisibility(true);
        return floatText;
    }

    private void StartTimers()
    {
        ReviveTimer.Timer.Start(TIME_TO_REVIVE, false, Revive);
        UpdateTextTimer.Timer.Start(0.03f, true, UpdateFloatingText);
    }

    private void UpdateFloatingText()
    {
        FloatingTimer.SetText($"{Colors.GetStringColorOfPlayer(Player.Id + 1)}{ReviveTimer.Timer.Remaining.ToString("F2")}|r", TextTagHeight);
    }

    private void Revive()
    {
        try
        {
            var kitty = Globals.ALL_KITTIES[Player];
            var lastCheckpoint = Globals.SAFE_ZONES[kitty.CurrentSafeZone];
            var x = lastCheckpoint.Rect_.CenterX;
            var y = lastCheckpoint.Rect_.CenterY;
            kitty.ReviveKitty();
            kitty.Unit.SetPosition(x, y);
            if (Player.IsLocal) PanCameraToTimed(x, y, 0.00f);
            CameraUtil.RelockCamera(Player);
            Dispose();
        }
        catch (System.Exception e)
        {
            Logger.Warning($"Error in SoloDeathTimer.Revive: {e.Message}");
            Dispose();
        }
    }

    private void Dispose()
    {
        ReviveTimer.Dispose();
        UpdateTextTimer.Dispose();
        FloatingTimer.Dispose();
        FloatingTimer = null;
    }
}

File: src/Gamemodes/Solo/SoloDeathTimer.ts
import { Logger } from 'src/Events/Logger/Logger'
import { Globals } from 'src/Global/Globals'
import { CameraUtil } from 'src/Utility/CameraUtil'
import { ColorUtils } from 'src/Utility/Colors/ColorUtils'
import { AchesTimers, createAchesTimer } from 'src/Utility/MemoryHandler/AchesTimers'
import { MapPlayer, TextTag } from 'w3ts'

export class SoloDeathTimer {
    private TIME_TO_REVIVE = 6.0
    private TextTagHeight = 0.018
    private Y_OFFSET = 5.0
    public ReviveTimer: AchesTimers
    public UpdateTextTimer: AchesTimers
    public Player: MapPlayer
    public FloatingTimer: TextTag

    public constructor(player: MapPlayer) {
        this.Player = player
        this.ReviveTimer = createAchesTimer()
        this.UpdateTextTimer = createAchesTimer()
        this.FloatingTimer = this.CreateFloatingTimer()
        this.StartTimers()
    }

    private CreateFloatingTimer(): TextTag {
        const circle = Globals.ALL_CIRCLES.get(this.Player)
        const floatText = TextTag.create()
        if (!circle || !floatText) return floatText! // xd
        floatText.setPos(circle.Unit.x, circle.Unit.y - this.Y_OFFSET, 0)
        floatText.setVisible(true)
        return floatText
    }

    private StartTimers = () => {
        this.ReviveTimer.start(this.TIME_TO_REVIVE, false, this.Revive)
        this.UpdateTextTimer.start(0.03, true, this.UpdateFloatingText)
    }

    private UpdateFloatingText = () => {
        SetTextTagText(
            this.FloatingTimer.handle,
            `${ColorUtils.GetStringColorOfPlayer(this.Player.id + 1)}${this.ReviveTimer.remaining().toFixed(2)}|r`,
            this.TextTagHeight
        )
    }

    private Revive = () => {
        try {
            const kitty = Globals.ALL_KITTIES.get(this.Player)!
            const lastCheckpoint = Globals.SAFE_ZONES[kitty.CurrentSafeZone]
            const x = lastCheckpoint.Rectangle.centerX
            const y = lastCheckpoint.Rectangle.centerY
            kitty.ReviveKitty()
            kitty.Unit.setPosition(x, y)
            if (this.Player.isLocal()) PanCameraToTimed(x, y, 0.0)
            CameraUtil.RelockCamera(this.Player)
            this.Dispose()
        } catch (e) {
            Logger.Warning(`Error in SoloDeathTimer.Revive: ${e}`)
            this.Dispose()
        }
    }

    private Dispose = () => {
        this.ReviveTimer.dispose()
        this.UpdateTextTimer.dispose()
        this.FloatingTimer.destroy()
    }
}
